### PR TITLE
Vertica driver

### PR DESCRIPTION
**Pull request by erp12**
Implements #1337 

For now, this PR is just in case anyone at metabase is available to help get the tests working for the Vertica driver.

Vertica driver has been used with our vertica database, and appears to work fine in the metabase application, although it has not undergone any of the metabase tests yet.

Current issue is that the prepared statements used in the metabase tests do no appear to be working the same in `jdbc` for `Vertica` as the are in `h2`.

NOTE: To use the veritca driver, I needed to put [the jdbc drvier jar](https://my.vertica.com/download/vertica/client-drivers/) in the plugins folder.

### CODE REVIEW

File: src/metabase/driver/vertica.clj [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+  [column-name column-type]
+  ;; this is really, really simple right now.  if its postgres :json type then it's :json special-type
+  (when (= column-type :json)
+    :json))
+
+(def ^:const ssl-params
+  "Params to include in the JDBC connection spec for an SSL connection."
+  {:ssl        true
+   :sslmode    "require"
+   :sslfactory "org.postgresql.ssl.NonValidatingFactory"})  ; HACK Why enable SSL if we disable certificate validation?
```
Comment (camsaul) [inline src/metabase/driver/vertica.clj]: probably doesn't make a ton of sense for the Vertica driver 😉

Comment (erp12) [inline src/metabase/driver/vertica.clj]: Good catch! The way I got the Vertica driver is by making a duplicate of the Postgres driver and debugging until it started working. I would guess there are a few things in the Vertica driver that don't need to be there, like this.

I removed this and a few other references to postgres stuff.

**Comment by erp12** [comment]
I have an update on prepared statements. 

I started up a local vertica db in docker and loaded in a table. When I run
`select * from customers;` in the `vsql` client I get this:

```
 CustID |                     Last_Name                      |                     First_Name                     |                       Email                        | Phone_Number
--------+----------------------------------------------------+----------------------------------------------------+----------------------------------------------------+--------------
      1 | Allen                                              | Anna                                               | aang@example.com                                   | 123-456-7890
      2 | Brown                                              | Bill                                               | b.brown@example.com                                | 555-444-3333
      3 | Chu                                                | Cindy                                              | cindy@example.com                                  | 555-867-5309
      4 | Dodd                                               | Don                                                | d.d@example.com                                    | 555-555-1212
      5 | Estavez                                            | Eric                                               | e.estavez@example.com                              | 781-555-0000 
(5 rows)
```

I then opened a fresh clojure project and used `clojure.java.jdbc` with the vertica jar included using the `lein-localrepo` plugin and ran this:
```
(execute! db ["INSERT INTO customers (Last_Name, First_Name, Email, Phone_Number) VALUES (?, ?, ?, ?)"
              "Fred" "Fanning" "ff@example.com" "978-645-3421"])
```

The result is almost exactly what we want/expect.
```
 select * from customers;
 CustID |                     Last_Name                      |                     First_Name                     |                       Email                        | Phone_Number
--------+----------------------------------------------------+----------------------------------------------------+----------------------------------------------------+--------------
      1 | Allen                                              | Anna                                               | aang@example.com                                   | 123-456-7890
      2 | Brown                                              | Bill                                               | b.brown@example.com                                | 555-444-3333
      3 | Chu                                                | Cindy                                              | cindy@example.com                                  | 555-867-5309
      4 | Dodd                                               | Don                                                | d.d@example.com                                    | 555-555-1212
      5 | Estavez                                            | Eric                                               | e.estavez@example.com                              | 781-555-0000
        | Fred                                               | Fanning                                            | ff@example.com                                     | 978-645-3421
(6 rows)
```
The `CustID` column is of type `AUTO_INCREMENT` which is vertica's version of Postgres's `SERIAL` type. I expected it to automatically know to put a `6` in that column but otherwise the prepared statement worked.

I then reset the db to be the initial dataset, and ran the follow prepared statement:
```
(execute! db ["INSERT INTO customers (Last_Name, First_Name, Email, Phone_Number) VALUES (?, ?, ?, ?), (?, ?, ?, ?)"
              "Fred" "Fanning" "ff@example.com" "978-645-3421"
              "Geff" "Giant" "gg@example.com" "482-381-3938"])
```

Now the issue shows itself! The response is:
```
SyntaxErrorException [Vertica][VJDBC](4856) ERROR: Syntax error at or near ","  com.vertica.util.ServerErrorData.buildException (:-1)
```

This suggested to me that `load-data-one-at-a-time-parallel!` would likely be the solution, and it appears to have fixed the issue.

I am now getting
```
[Vertica][VJDBC](2444) ERROR: Cannot insert into or update IDENTITY/AUTO_INCREMENT column "id"
```
In many of the insert statements. Its clear that the cause of this is that metabase is trying to assign an integer to "id" while Vertica wants to handle that because the "id" column is of type `AUTO_INCREMENT`. I haven't had a chance to poke around for a solution yet, but I have lots of time set aside for it tomorrow.

**Comment by camsaul** [comment]
@erp12 `load-data-one-at-a-time-parallel!` specifies the IDs in the `INSERT` statements so the data loaded into the DB will always have the correct IDs no matter what order the `INSERTS` actually happen. Since it sounds like Vertica doesn't like that you should just use `load-data-one-at-a-time!` instead. It's slower but since it the `INSERT`s are guaranteed to happen in a certain order we don't need to specify the IDs ourselves; the implementation skips the step which should resolve your problems

**Comment by erp12** [comment]
@camsaul Thanks! That was very helpful. It turns out that `AUTO_INCREMENT` does assign unique, increasing, integer ids to each row but due to Vertica's distributed nature they aren't always sequential. In fact, the default is to skip ahead by 250000 every time a new session is started, which is bad news for `clojure.java.jdbc`.

In other words, if I connect to Vertica, create a new table, and insert 3 rows with `AUTO_INCREMENT` ids, the ids will be 1, 2, and 3.
If I disconnect from Vertica, reconnect, and insert 3 new rows into the same table, their ids will be 250001, 250002, and 250003.
From what I have read, the reason for this is that Vertica can't be sure what ids have changed because stuff might be happening on various nodes... or some other reason relating to being a distributed database.

I ended up setting `:pk-sql-type` to `(constantly "INTEGER")` instead of `(constantly "AUTO_INCREMENT")` and using `load-data-one-at-a-time-parallel!`. Now ids aren't `AUTO_INCREMENT`s anymore and the ids are set by metabase.

**Comment by camsaul** [comment]
@erp12 I don't think that should be an issue since the test code reuses the same connection for loading all the data

**Comment by erp12** [comment]
@camsaul I am not very familiar with `clojure.java.jdbc` but in my testing outside of metabase, it was assigning ids every by incrementing 250000 with every call to `jdbc/execute!`. 

Forgive me if I am getting some of the terminology wrong here, but isn't it true that most `clojure.java.jdbc` functions take a map that describes the how to connect to the db, but then the actual connection is made by jdbc in every call?

**Comment by camsaul** [comment]
@erp12 the map can contain a reference to an existing open connection (which itself can even be a connection pool proxy connection, which is what we use in most places). `clojure.java.jdbc` is flexible in that it can either open the connection itself or reüse an existing one 

That being said if you're seeing the numbers increment by that much every time then switching to `integer` is probably your best bet. Did that work?

**Comment by erp12** [comment]
@camsaul perhaps I am using the word "connection" wrong here... 
Yes, switching to `integer` worked well. Also, now the data loads a bit faster because we can use `load-data-one-at-a-time-parallel!`

**Comment by camsaul** [comment]
@erp12 so are tests passing for you now locally?

**Comment by erp12** [comment]
@camsaul Not quite 😄  but I don't appear to be getting errors related to loading the test data, and the number of errors I am seeing has dropped a lot.

I am getting a few errors now in `query_processor_test.clj`. I don't know much about the different kinds of databases, and the `expect-with-non-timeseries-dbs` functions are kind of a mystery to me.

It seems to me like it is running a test, and comparing against an expected result. Different database engines have different expected results. But what do the expected results actually mean?

For example, I initially had an error with [this `expect-with-non-timeseries-dbs`](https://github.com/metabase/metabase/blob/7f5072cb51cf307dc29b14ab482e3db97abe78fd/test/metabase/query_processor_test.clj#L1418-L1455).

The error said this was the expected:
```
[["2015-06-01T03:31:00.000Z" 1]
     ["2015-06-01T09:06:00.000Z" 1]
     ["2015-06-01T10:23:00.000Z" 1]
     ["2015-06-01T11:55:00.000Z" 1]
     ["2015-06-01T14:04:00.000Z" 1]
     ["2015-06-01T14:19:00.000Z" 1]
     ["2015-06-01T19:13:00.000Z" 1]
     ["2015-06-01T22:37:00.000Z" 1]
     ["2015-06-02T01:20:00.000Z" 1]
     ["2015-06-02T04:11:00.000Z" 1]]
```
which I recognized as the `:else` in the `cond`. The error then said that this is what the actual result was:
```
    [["2015-06-01T10:31:00.000Z" 1]
     ["2015-06-01T16:06:00.000Z" 1]
     ["2015-06-01T17:23:00.000Z" 1]
     ["2015-06-01T18:55:00.000Z" 1]
     ["2015-06-01T21:04:00.000Z" 1]
     ["2015-06-01T21:19:00.000Z" 1]
     ["2015-06-02T02:13:00.000Z" 1]
     ["2015-06-02T05:37:00.000Z" 1]
     ["2015-06-02T08:20:00.000Z" 1]
     ["2015-06-02T11:11:00.000Z" 1]]
```
which I recognized as the `(contains? #{:redshift :sqlserver :bigquery :mongo :postgres :h2 :oracle} *engine*)` in the 'cond'. I figured I would try throwing `:vertica` into the condition and re-run the tests, and I no longer get the error. I don't know if that was the right thing to do... but it seems to me that Vertica has the same behavior as `postgres`, `h2`, `oracle`, etc for that test so metabase should expect the same result.

A more puzzling example is the next error I get from [this `expect-with-non-timeseries-dbs`](https://github.com/metabase/metabase/blob/7f5072cb51cf307dc29b14ab482e3db97abe78fd/test/metabase/query_processor_test.clj#L1635-L1646).

Error:
```
expected:
 [[23 46] [24 47] [25 40] [26 60] [27 7]]
 
was:
 [[23 38] [24 50] [25 42] [26 55] [27 15]]
```

This time I don't see any engine that expects the result `[[23 38] [24 50] [25 42] [26 55] [27 15]]`. I am not really sure how to go about fixing this because I am not sure what this vector of integer pairs is supposed to be. Any thoughts?

**Comment by camsaul** [comment]
@erp12 any luck with that? Is there still much to do for this Vertica driver?

**Comment by erp12** [comment]
I think I only have this last  `:week-of-year` issue left to fix. 

```
Ran 1189 tests containing 1189 assertions in 97335 msecs
1 failures, 0 errors.
Error encountered performing task 'expectations' with profile(s): 'base,system,user,provided,dev,expectations'
```

However, I do see quite a few warnings. I am planning on seeing what I can do about them after all the errors are solved.

I have spent some time reading through the `metabase.util.honeysql-extensions` functions to try and get a sense of what the current value for `:week-of-year` is doing for the various existing metabase drivers. Unfortunately, I have never used `honeysql` so its taking me a while to fully understand how this:
```
#sql/call [:to_timestamp #sql/call [:/ :public.sad_toucan_incidents_incidents.timestamp 1000]]
```
becomes a query that is run agains the database.

**Comment by camsaul** [comment]
@erp12 I'd recommend just opening up a REPL and passing the HoneySQL code to `hsql/format` to see what SQL gets produced. In the example you have above we get:

```clojure
(hsql/call :to_timestamp (hsql/call :/ :public.sad_toucan_incidents.timestamp 1000))

;; -> "to_timestamp((public.sad_toucan_incidents.timestamp / 1000))"
```

If you can give me a sense of what SQL you think should be produced for `:week-of-year` I can help with the HoneySQL version. I'd also recommend taking a look at how the various other SQL drivers handle `:week-of-year`; we have 7 or 8 of them now so one of them is probably very close to what you want.

**Comment by erp12** [comment]
Thanks for the info! I tried a few different things for `:week-of-year` and found one that fixed the issue. So...

```
Ran 1563 tests containing 1563 assertions in 101538 msecs
0 failures, 0 errors.
``` 
😄

I pulled the most recent metabase/master to confirm that all the tests were still passing and the _tests_ are still passing, but I am now seeing a new error about failing to connect to a local `postgres` database:

```
ERROR metabase.driver :: Failed to connect to database: java.sql.SQLException: No suitable driver found for jdbc:postgresql://localhost:/mem:test-data;USER=GUEST;PASSWORD=guest
ERROR metabase.driver :: Failed to connect to database: org.postgresql.util.PSQLException: Connection refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
ERROR metabase.driver :: Failed to connect to database: org.postgresql.util.PSQLException: Protocol error.  Session setup failed.
```

Has something changed about the tests and we will now need a local postgres database available? If so, I can easily set that up.

**Comment by camsaul** [comment]
@erp12 We have some tests that try to connect to a Postgres DB that doesn't exist to make sure it fails, and they output those warning messages. So you can safely ignore those.

Great work on getting the Vertica driver going! 🎉 

The next step is to figure out how to get tests going on CI. It looks like HP offers [an AMI that's preconfigured for running Vertica on AWS](https://my.vertica.com/download/vertica/amazon-web-services/). It also looks like the offer a [community edition](https://my.vertica.com/download/vertica/community-edition/) that runs on Ubuntu and other Linuxes.

How are you running Vertica?

**Comment by erp12** [comment]
@camsaul Okay thats good news about the `postgres` database. I am attaching [the full result of running the tests](https://github.com/metabase/metabase/files/512311/metabase_test.txt) for you to review to make sure that everything looks normal to you. Let me know if any of these warning are things I should address.

I am running vertica in docker in the way **twneale** suggested [here](https://github.com/metabase/metabase/issues/1337). What we use for development is very similar to [this](https://github.com/sumitchawla/docker-vertica/blob/master/Dockerfile). I can get more information about this approach if you would like.

**Comment by camsaul** [comment]
@erp12, I must be missing something when trying to run the tests.

I started the Docker image you pointed me towards:

```bash
$ docker run -p 5433:5433 sumitchawla/vertica
[...]
Database docker started successfully
Vertica is now running
```

and put the Vertica JDBC driver in my `plugins/` directory, and run the tests:

```bash
MB_DB_TYPE=h2 ENGINES=h2,vertica lein test
[...]
  act-msg: exception in actual: (metabase.test.data.datasets/do-with-engine :vertica actual-875170329)
    threw: class java.sql.SQLNonTransientConnectionException - [Vertica][VJDBC](100176) Failed to connect to host localhost on port 5433. Reason: Failed to establish a connection to the primary server or any backup address.
```

**Comment by camsaul** [comment]
@erp12 let me know if you can explain what I'm doing wrong here, we're super close to shipping our long-awaited much-hyped 0.20 release and if I can get Vertica tests working on CI there's a good chance we can include it in the release

**Comment by camsaul** [comment]
@erp12 nevermind, I was running Docker on a VM so I was using localhost when I shouldn't have been. I'm seeing this error message now:

```
java.sql.SQLNonTransientConnectionException: [Vertica][VJDBC](5273) FATAL: Unsupported frontend protocol 3.6: server supports 3.0 to 3.5
```

I have the latest 8.0 JDBC driver, I'm guessing maybe the Docker image has an older version? Do you know anything about that?

**Comment by camsaul** [comment]
Yup, [sounds like the new Vertica 8.x JDBC driver doesn't work for Vertica 7.x](https://community.dev.hpe.com/t5/Vertica-Forum/Vertica-JDBC-client-issue/td-p/219207)

Who does that?

@erp12 it sounds like you must be using 7.x, right? I'm hesitant to ship a driver that only works with older versions of Vertica.

**Comment by erp12** [comment]
@camsaul Just getting caught up on this. Yes we use Vertica 7.2.2.  That is what I was developing/testing with. That is very odd that their drivers aren't backwards compatible.

**Comment by camsaul** [comment]
I think this might not be too much of an issue as long as people pair the right JDBC driver with whatever version of Vertica they have. 💔

**Comment by camsaul** [comment]
@erp12 do you know of a docker image running 7.2.2? The one you pointed me towards is 7.1.1. I tried the JDBC driver version 7.2.3 and that doesn't work with 7.1.1 either

**Comment by erp12** [comment]
@camsaul I could look into getting a docker image for vertica 8.0 and re-running the tests with the 8.x jdbc har, but I am not confident I would be able to re-distribute it.

I wonder if the free community edition of vertica can be paired with docker?

RE 7.2.2: Will [this](https://github.com/jbfavre/docker-vertica) work?

**Comment by camsaul** [comment]
@erp12 I actually got DB version 7.1.1 working with JDBC driver version 7.1.2

**Comment by camsaul** [comment]
How are you running your production Vertica server? I tried spinning up an Elastic Beanstalk app using that Dockerfile and a copy of the Community Edition 8.0.0 package but it's complaining that it doesn't like the instance size:

```
  AWS Detected. Using AWS defaults.
      AWS Default: --data-dir was not specified, using default: /vertica/data
      AWS Default: --point-to-point was not specified,  enabling point-to-point spread communication by default while on AWS
  
  Error: t2.small is not a valid AWS instance type.
  Hint: Valid types are: c4.4xlarge,c4.8xlarge,c3.4xlarge,c3.8xlarge,r3.4xlarge,r3.8xlarge,d2.4xlarge,d2.8xlarge,i2.4xlarge,i2.8xlarge,m4.4xlarge,m4.10xlarge
```

I think running a `4xlarge` instance is a little out of our price range so trying to figure out a way to work around this.

**Comment by erp12** [comment]
@camsaul Thats great! Hopefully the tests pass regardless of jdbc driver versions... If not, I beleive that we (as enterprise users) have access to multiple previous versions of everything. Its possible I can run tests for other versions.

Regarding our production server, I don't have many answers other than we have our own Vertica cluster... possibly on AWS, but I don't think so. @twneale will know.

Unfortunately, I don't have a very strong background in databases, but other members of [our team](https://github.com/massmutual) have been working with vertica for a while now. I can ask around to get more info if needed.

**Comment by camsaul** [comment]
@erp12 great news, I've ran your tests successfully locally against 7.1 and everything passed. Can you sign our [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)? 

As soon as I figure out how to get Vertica running somewhere in the cloud this is ready to be merged

**Comment by erp12** [comment]
@camsaul That great! I am glad it worked out.
Just signed the CLA.

**Comment by camsaul** [comment]
@erp12 I'm still in the process of getting this set up to run on CI. I've opened a new PR with the commits you made plus my work towards CI integration at #3468, so I'm going to go ahead and close this out in favor of that one.

Keep an eye it. We're planning on releasing the Vertica driver as part of our 0.21.0 release in a month or so.

**Comment by erp12** [comment]
Sounds great, Thanks!

**Pull request closed without merging**